sonar-db/src/main/java/org/sonar/db/permission/UserPermissionDao.java

```diff
@@ -0,0 +1,98 @@
+ */
+package org.sonar.db.permission;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.ibatis.session.RowBounds;
+import org.sonar.db.Dao;
+import org.sonar.db.DatabaseUtils;
+import org.sonar.db.DbSession;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static org.apache.commons.lang.StringUtils.isNotEmpty;
+import static org.sonar.db.DatabaseUtils.executeLargeInputs;
+
+public class UserPermissionDao implements Dao {
+
+  /**
+   * @see UserPermissionMapper#selectByQuery(PermissionQuery, Collection, RowBounds)
+   */
+  public List<ExtendedUserPermissionDto> select(DbSession dbSession, PermissionQuery query, @Nullable Collection<String> userLogins) {
+    if (userLogins != null) {
+      if (userLogins.isEmpty()) {
+        return emptyList();
+      }
+      checkArgument(userLogins.size() <= DatabaseUtils.PARTITION_SIZE_FOR_ORACLE, "Maximum 1'000 users are accepted");
+    }
+
+    RowBounds rowBounds = new RowBounds(query.getPageOffset(), query.getPageSize());
+    return mapper(dbSession).selectByQuery(query, userLogins, rowBounds);
+  }
+
+  /**
+   * @see UserPermissionMapper#countUsersByQuery(PermissionQuery, Collection)
+   */
+  public int countUsers(DbSession dbSession, PermissionQuery query) {
+    return mapper(dbSession).countUsersByQuery(query, null);
+  }
+
+  /**
+   * Count the number of users per permission for a given list of projects
+   *
+   * @param projectIds a non-null list of project ids to filter on. If empty then an empty list is returned.
+   */
+  public List<UserCountPerProjectPermission> countUsersByProjectPermission(DbSession dbSession, Collection<Long> projectIds) {
+    return executeLargeInputs(projectIds, mapper(dbSession)::countUsersByProjectPermission);
+  }
+
+  public void insert(DbSession dbSession, UserPermissionDto dto) {
+    mapper(dbSession).insert(dto);
+  }
+
+  /**
+   * Delete permissions for a user, permissions for a project, or a mix of them. In all cases
+   * scope can be restricted to a specified permission.
+   * @see UserPermissionMapper#delete(String, String, String)
+   */
+  public void delete(DbSession dbSession, @Nullable String login, @Nullable String projectUuid, @Nullable String permission) {
+    checkArgument(isNotEmpty(login) || isNotEmpty(projectUuid), "At least one of login or project must be set");
+    mapper(dbSession).delete(login, projectUuid, permission);
+  }
+
+  // shortcuts for migration
+  public List<String> selectUserPermissions(DbSession dbSession, String userLogin, @Nullable String projectUuid) {
+    PermissionQuery query = PermissionQuery.builder()
+      .withAtLeastOnePermission()
+      .setComponentUuid(projectUuid)
+      .build();
+    return select(dbSession, query, asList(userLogin)).stream()
+      .map(p -> p.getPermission())
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Replace this lambda with a method reference. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)